src/Control/Monad/Freer/Resource.hs

```diff
@@ -0,0 +1,262 @@
+{-# LANGUAGE ConstraintKinds                    #-}
+{-# LANGUAGE DataKinds                          #-}
+{-# LANGUAGE FlexibleContexts                   #-}
+{-# LANGUAGE FlexibleInstances                  #-}
+{-# LANGUAGE GADTs                              #-}
+{-# LANGUAGE MultiParamTypeClasses              #-}
+{-# LANGUAGE PolyKinds                          #-}
+{-# LANGUAGE RankNTypes                         #-}
+{-# LANGUAGE ScopedTypeVariables                #-}
+{-# LANGUAGE TypeApplications                   #-}
+{-# LANGUAGE TypeFamilies                       #-}
+{-# LANGUAGE TypeOperators                      #-}
+{-# LANGUAGE UndecidableInstances               #-}
+{-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
+
+module Control.Monad.Freer.Resource
+  ( SafeForRegion
+  , Resource (..)
+  , HiddenResource
+  , ResourceCtor
+  , catchNothing
+  , acquire
+  , acquire'
+  , handleRegionRelay
+  , fileHandleRegion
+  , give
+  , thisRegion
+  , parentRegion
+  , gparentRegion
+  ) where
+
+import Control.Exception (SomeException, catch)
+import Control.Monad.Freer
+import Control.Monad.Freer.Exception (Exc (..), throwError)
+import Control.Monad.Freer.Internal (Union, Eff (..), qComp, tsingleton, prj, decomp)
+import Data.Bool (bool)
+import Data.List (delete)
+import Data.Proxy (Proxy (..))
+import GHC.TypeLits (Nat, type (+), type (-), type (<=))
+import System.IO
+
+
+------------------------------------------------------------------------------
+-- | Instances of this typeclass describe which effects 'r' are allowed to be
+-- present inside of a region for 'res'.
+class SafeForRegion res (r :: [* -> *])
+
+instance SafeForRegion res '[]
+instance SafeForRegion res r => SafeForRegion res (RegionEff res s ': r)
+
+instance SafeForRegion Handle r => SafeForRegion Handle (Exc SomeException ': r)
```

Comment:
acquiring/releasing handles can throw `SomeException`s, so in order to reify that into our stack, we need to identify it here